### PR TITLE
8312610: GenShen: Old generation available is unintentionally restricted by mutator's available memory

### PR DESCRIPTION
**Pull request by earthling-amzn**
Only young gen allocations should be restricted to memory in the mutator's view of the freeset.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8312610](https://bugs.openjdk.org/browse/JDK-8312610): GenShen: Old generation available is unintentionally restricted by mutator's available memory (**Bug** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [0df69803](https://git.openjdk.org/shenandoah/pull/301/files/0df698036c053244bbeb9ab2fa38b3c386e332ed)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Committer) ⚠️ Review applies to [0df69803](https://git.openjdk.org/shenandoah/pull/301/files/0df698036c053244bbeb9ab2fa38b3c386e332ed)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/301/head:pull/301` \
`$ git checkout pull/301`

Update a local copy of the PR: \
`$ git checkout pull/301` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/301/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 301`

View PR using the GUI difftool: \
`$ git pr show -t 301`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/301.diff">https://git.openjdk.org/shenandoah/pull/301.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/301#issuecomment-1648120101)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahYoungGeneration.cpp [hunk 1]
```diff
@@ -95,3 +95,15 @@ void ShenandoahYoungGeneration::add_collection_time(double time_seconds) {
     ShenandoahGeneration::add_collection_time(time_seconds);
   }
 }
+
+size_t ShenandoahYoungGeneration::available() const {
+  // The collector reserve may eat into what the mutator is allowed to use. Make sure we are looking
+  // at what is available to the mutator when reporting how much memory is avilable.
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahYoungGeneration.cpp:101]: small typo here: available

**Review by kdnilsen — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahGlobalGeneration.cpp [hunk 2]
```diff
@@ -58,6 +58,14 @@ size_t ShenandoahGlobalGeneration::available() const {
   return ShenandoahHeap::heap()->free_set()->available();
 }
 
+size_t ShenandoahGlobalGeneration::soft_available() const {
+  size_t available = this->available();
+
+  // Make sure the code below treats available without the soft tail.
+  size_t soft_tail = max_capacity() - soft_max_capacity();
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahGlobalGeneration.cpp:65]: Would it be useful to assert here that `max_capacity() >= soft_max_capacity()` ?

**Review by ysramakrishna — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahYoungGeneration.cpp [hunk 3]
```diff
@@ -95,3 +95,15 @@ void ShenandoahYoungGeneration::add_collection_time(double time_seconds) {
...
+  size_t available = this->ShenandoahGeneration::available();
+  return MIN2(available, ShenandoahHeap::heap()->free_set()->available());
+}
+
+size_t ShenandoahYoungGeneration::soft_available() const {
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahYoungGeneration.cpp:106]: No - this is a bug. Good catch!

File: src/hotspot/share/gc/shenandoah/shenandoahGlobalGeneration.cpp (hunk 2 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahGlobalGeneration.cpp:65]: Yes, better than computing an underflowed valued.

**Pull request closed without merging**